pallets/dkg-metadata/src/lib.rs

```diff
@@ -1574,6 +1492,126 @@ impl<T: Config> Pallet<T> {
+	/// that need a fallback mechanism when the DKG is not available or needs to be fixed or
+	/// changed.
+	fn create_voter_set_data() -> ([u8; 32], u64, u32) {
+		let voters = Self::next_authorities();
+		// Merkleize the new voter set
+		let voter_set_merkle_root = Self::get_voter_set_tree_root(&voters);
+		// Get average session length
+		let average_session_length_in_blocks: u64 =
+			T::NextSessionRotation::average_session_length().try_into().unwrap_or_default();
+
+		let average_millisecs_per_block: u64 =
+			<T as pallet_timestamp::Config>::MinimumPeriod::get()
+				.saturating_mul(2u32.into())
+				.try_into()
+				.unwrap_or_default();
+
+		let average_session_length_in_millisecs =
+			average_session_length_in_blocks.saturating_mul(average_millisecs_per_block);
+
+		(voter_set_merkle_root, average_session_length_in_millisecs, voters.len() as u32)
+	}
+
+	/// Returns the leaves of the voter set merkle tree.
+	///
+	/// It is expected that the size of the returned vector is a power of 2.
+	pub fn pre_process_for_merkleize(voters: &[T::DKGId]) -> Vec<[u8; 32]> {
+		let height = Self::get_voter_set_tree_height(voters.len());
+		// Hash the external accounts into 32 byte chunks to form the base layer of the merkle tree
+		let mut base_layer: Vec<[u8; 32]> = voters
+			.iter()
+			.map(|account| account.to_raw_vec())
+			.map(|account| keccak_256(&account))
+			.collect();
+		// Pad base_layer to have length 2^height
+		let two = 2;
+		while base_layer.len() != two.saturating_pow(height.try_into().unwrap_or_default()) {
+			base_layer.push(keccak_256(&[0u8]));
+		}
+		base_layer
+	}
+
+	/// Computes the next layer of the merkle tree by hashing the previous layer.
+	pub fn next_layer(curr_layer: Vec<[u8; 32]>) -> Vec<[u8; 32]> {
+		let mut layer_above: Vec<[u8; 32]> = Vec::new();
+		let mut index = 0;
+		while index < curr_layer.len() {
+			let mut input_to_hash_as_vec: Vec<u8> = curr_layer[index].to_vec();
+			input_to_hash_as_vec.extend_from_slice(&curr_layer[index + 1][..]);
+			let input_to_hash_as_slice = &input_to_hash_as_vec[..];
+			layer_above.push(keccak_256(input_to_hash_as_slice));
+			index += 2;
+		}
+		layer_above
+	}
+
+	// Returns the minimal height of the voter set Merkle tree
+	// Right now this takes O(log(size of voter set)) time but can likely be reduced
+	pub fn get_voter_set_tree_height(voter_count: usize) -> u32 {
+		if voter_count == 1 {
+			1
+		} else {
+			let two: u32 = 2;
+			let mut h = 0;
+			while two.saturating_pow(h) < voter_count as u32 {
+				h += 1;
+			}
+			h
+		}
+	}
+
+	/// Computes the merkle root of the voter set tree
+	pub fn get_voter_set_tree_root(voters: &[T::DKGId]) -> [u8; 32] {
+		let mut curr_layer = Self::pre_process_for_merkleize(voters);
+		let mut height = Self::get_voter_set_tree_height(voters.len());
+		while height > 0 {
+			curr_layer = Self::next_layer(curr_layer);
+			height -= 1;
+		}
+
+		let mut root = [0u8; 32];
+		root.copy_from_slice(&curr_layer[0][..]);
+		root
+	}
```

Comment:
Could we use a proper merkle tree implementation instead of rolling out our custom code here?

Reply:
I support this as a separate task.